python/TestHarness/TestHarness.py

```diff
@@ -235,87 +242,140 @@ def launchTest(self, find_only, file, dirpath, testers):
+                # testers as a whole get repeatedly printed as we iterate over each PBS test.
+
+                # TODO: Try and figure out a way to print these skipped tests only once when processing
+                # PBS batch history.
+                if should_run or self.options.processingPBS:
+                    command = tester.getCommand(self.options)
+
+                    # Handle PBS processing
+                    # TODO: figure out how to allow getCommand do this
+                    if self.options.pbs:
+                        # Get qstat results
+                        if self.options.processingPBS:
+                            if tester.specs['test_name'] == pbs_job['test_name']:
+                                command = 'qstat -x -f %s' % (pbs_job['id'])
+                            else:
+                                continue
+
+                        # Create qsub script.
+                        else:
+                            # Override command with the qsub command needed to launch a PBS job
+                            command = self.createQsubFile(command, dirpath, tester)
+
+                    # This method spawns another process and allows this loop to continue looking for tests
+                    # RunParallel will call self.testOutputAndFinish when the test has completed running
+                    # This method will block when the maximum allowed parallel processes are running
+                    if self.options.dry_run:
                         self.handleTestStatus(tester)
-                self.runner.jobSkipped(tester.parameters()['test_name'])
+                    else:
+                        self.runner.run(tester, command)
+                else: # This job is skipped - notify the runner
+                    status = tester.getStatus()
+                    if status != tester.bucket_silent: # SILENT occurs when a user is using --re options
+                        if (self.options.report_skipped and status == tester.bucket_skip) \
+                           or status == tester.bucket_skip:
+                            self.handleTestStatus(tester)
+                        elif status == tester.bucket_deleted and self.options.extra_info:
+                            self.handleTestStatus(tester)
+                    self.runner.jobSkipped(tester.parameters()['test_name'])
+
+    # Create the PBS qsub script, perform directory/file operations
+    # TODO: Because we need to open this file a second time to modify
+    # any would-be prereq attribute, should we refactor this, and move
+    # this routine to RunParallel (if possible)?
+    # see addJobID in RunParallel for clarification.
+    def createQsubFile(self, command, dirpath, tester):
+        pbs_job_dir = os.path.join(dirpath, 'job_' + self.options.pbs)
+        if not os.path.exists(pbs_job_dir):
+            os.mkdir(pbs_job_dir)
+        os.chdir(pbs_job_dir)
```

Comment:
Thats right! No more serialization. The directory created will be a unique name based on the arguments giving to --pbs. This will always work, because if you are supplying --pbs with a name already taken, then what are trying to do is test completed results based on a past run instead (processingPBS=True).